### DISCUSSION TITLE
Stop actor thread completely after timeout

### DISCUSSION THREAD

**Discussion by udit-pandey** [jodal/pykka#202] [Q&A]
I was exploring pykka for one of my use cases in following manner:

    import threading
    import time
    
    import pykka
    
    
    class AnActor(pykka.ThreadingActor):
        
        def func(self):
            i = 0
            while i < 4:
                i += 1
                time.sleep(5)  # Block a bit to make it realistic
                print(f"{i}: printing after 5 secs")
    
    
    def log(msg):
        thread_name = threading.current_thread().name
        print(f"{thread_name}: {msg}")
    
    
    try:
        actor = AnActor.start().proxy()
        log("calling AnActor.func() ...")
        result = actor.func()  # Does not block, returns a future
        log("printing result ... (blocking)")
        log(result.get(timeout=1))  # Blocks until ready
    except Exception as e:
        log(e)
        actor.stop()
        raise e

Output:

    MainThread: calling AnActor.func() ...
    MainThread: printing result ... (blocking)
    Traceback (most recent call last):
      File "./venv/lib/python3.8/site-packages/pykka/_threading.py", line 48, in get
        self._result = self._queue.get(True, timeout)
      File "/usr/lib/python3.8/queue.py", line 178, in get
        raise Empty
    _queue.Empty
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "./anactor.py", line 32, in <module>
        raise e
      File "./anactor.py", line 28, in <module>
        log(result.get(timeout=1))  # Blocks until ready
      File "./venv/lib/python3.8/site-packages/pykka/_threading.py", line 59, in get
        raise Timeout(f"{timeout} seconds")
    pykka._exceptions.Timeout: 1 seconds
    MainThread: 1 seconds
    1: printing after 5 secs
    2: printing after 5 secs
    3: printing after 5 secs
    4: printing after 5 secs
    
    Process finished with exit code 1

Basically, I want to trigger AnActor's func method but if the execution time exceeds timeout, the thread must be stopped completely. Any exception in actor execution must be raised too. 
In above example, timeout is happening but the program exit is not happening until the execution of actor is complete. Can such a use case be implemented through pykka?

Python v3.8.10
Pykka v3.1.1

**Comment by jodal** [comment]
The future object doesn't have any special access to the running actor, so when you block on the future it just waits for a return value message to come across a message queue.

Similarly, the running actor have no idea if there's anyone on the other end of the future waiting for a response, or if the consumer times out while waiting.

I don't know exactly what you're trying to do, but what might help if you want to be able to stop the actor earlier on some condition is to split up `func()` so that it does one step of its work and then calls itself through `self.actor_ref.proxy()` to do the next step of the work. That way, a request to stop the actor can be processed in-between two steps of the work.

```
def func(self, i=1):
    time.sleep(5) # Block a bit to make it realistic
    print(f"{i}: printing after 5 secs")
    if i < 4:
        self.actor_ref.proxy().func(i=i + 1)
```

**Comment by udit-pandey** [comment]
Thanks for the reply!
I want to build an actor that does some task but if the execution time of that task exceeds a certain limit, then the actor must be stopped. The task may or may not have an output that should be awaited. The task is not splittable as of now. Is this possible with what we have currently?

**Reply by jodal** [reply]
There's no current feature in Pykka fitting to watch and potentially cancel a long-running task.
